src/shared/code-paste-handler-plugin.ts

```diff
@@ -0,0 +1,98 @@
+import { Plugin } from "prosemirror-state";
+import { EditorView } from "prosemirror-view";
+import { Slice, Node } from "prosemirror-model";
+import { EditorType as EditorTypeName } from "./view";
+import type { EditorType } from "./view";
+
+/**
+ * Detects if code was pasted into the document and returns the text if true
+ * @param clipboardData The clipboardData from the ClipboardEvent
+ */
+export function getDetectedCode(clipboardData: DataTransfer): string | null {
+    // if we're loading a whole document, don't false positive if there's more than just code
+    const htmlContent = clipboardData.getData("text/html");
+    if (htmlContent && htmlContent.includes("<code>")) {
+        const allContent = new DOMParser().parseFromString(
+            htmlContent,
+            "text/html"
+        );
+        const codeBlock = allContent.querySelector("code");
+
+        return allContent.body.textContent.trim() !== codeBlock.textContent
+            ? null
+            : codeBlock.textContent;
+    }
+
+    const textContent = clipboardData.getData("text/plain");
+
+    if (!textContent) {
+        return null;
+    }
+
+    // TODO how to reliably detect if a string is code?
+
+    // TODO add more support?
+    // check if there's ide specific paste data present
+    if (clipboardData.getData("vscode-editor-data")) {
+        // TODO parse data for language?
+        return textContent;
+    }
+
+    // no ide detected, try detecting leading indentation
+    // true if any line starts with: 2+ space characters, 1 tab character
+    if (/^([ ]{2,}|\t)/m.test(textContent)) {
+        return textContent;
+    }
+
+    return null;
+}
+
+/** Plugin that auto-detects if code was pasted and handles it specifically */
+export const codePasteHandler = (editorType: EditorType) =>
+    new Plugin({
+        props: {
+            handlePaste(view: EditorView, event: ClipboardEvent, slice: Slice) {
+                let codeData: string;
+
+                const isMarkdown = editorType === EditorTypeName.Commonmark;
+                // if the schema parser already detected a code block, just use that
+                if (
+                    slice.content.childCount === 1 &&
+                    slice.content.child(0).type.name === "code_block"
+                ) {
+                    codeData = slice.content.child(0).textContent;
+                } else {
+                    codeData = getDetectedCode(event.clipboardData);
+                }
+
+                if (!codeData) {
+                    return false;
+                }
+
+                // TODO can we do some basic formatting?
+
+                const selectedNode = view.state.selection.$from.node();
+
+                if (isMarkdown) {
+                    codeData = `\n\`\`\`\n${codeData}\n\`\`\``;
+                }
```

Comment:
This is the meat 🥩

Reply:
Also, those `\n` are not the best approach and I plan to clean this up by detecting if selection is on a new line or not.